dotCMS/src/main/java/com/dotcms/contenttype/business/ContentTypeAPIImpl.java

```diff
@@ -802,31 +803,118 @@ public List<ContentType> search(String condition, BaseContentType base, final St
   @Override
   public List<ContentType> search(final List<String> sites, final String condition, final BaseContentType base, final String orderBy, final int limit, final int offset)
           throws DotDataException {
+      return search(sites, condition, base, orderBy, limit, offset, null);
+  }
 
-    final List<ContentType> returnTypes = new ArrayList<>();
-    int rollingOffset = offset;
-    try {
-      while ((limit<0)||(returnTypes.size() < limit)) {
-        final List<String> resolvedSiteIds = HostUtil.resolveSiteIds(sites, this.user, this.respectFrontendRoles);
-        final List<ContentType> rawContentTypes = this.contentTypeFactory.search(resolvedSiteIds,
-                condition, base.getType(), orderBy, limit, rollingOffset);
-        if (rawContentTypes.isEmpty()) {
-          break;
-        }
-        returnTypes.addAll(this.perms.filterCollection(rawContentTypes, PermissionAPI.PERMISSION_READ, this.respectFrontendRoles, this.user));
-        if(returnTypes.size() >= limit || rawContentTypes.size()<limit) {
-          break;
-        }
-        rollingOffset += limit;
+  @CloseDBIfOpened
+  @Override
+  public List<ContentType> search(final List<String> sites, final String condition,
+          final BaseContentType base, final String orderBy, final int limit, final int offset,
+          final List<String> requestedContentTypes) throws DotDataException {
+
+      final List<ContentType> returnTypes = new ArrayList<>();
+      final Set<String> includedIds = new HashSet<>();
+
+      final List<String> resolvedSiteIds = HostUtil.resolveSiteIds(sites, this.user,
+              this.respectFrontendRoles);
+
+      try {
+          // Handle explicitly requested Content Types
+          if (requestedContentTypes != null && !requestedContentTypes.isEmpty()) {
+              final List<ContentType> includedContentTypes = this.contentTypeFactory.find(
+                      requestedContentTypes, null, offset, limit, orderBy);
+
+              final List<ContentType> authorizedIncluded = this.perms.filterCollection(
```

Comment:
why is the filterCollection method called only when the requestedContentTypes are sent?